OpenSim/Moco/MocoFrameOrientationConstraint.cpp

```diff
@@ -0,0 +1,109 @@
+    constructProperties();
+}
+
+MocoFrameOrientationConstraintPair::MocoFrameOrientationConstraintPair(
+    std::string frame1Path, std::string frame2Path,
+    double minimum_angle, double maximum_angle) {
+    constructProperties();
+    set_frame1_path(frame1Path);
+    set_frame2_path(frame2Path);
+    set_minimum_angle(minimum_angle);
+    set_maximum_angle(maximum_angle);
+}
+
+void MocoFrameOrientationConstraintPair::constructProperties() {
+    constructProperty_frame1_path("");
+    constructProperty_frame2_path("");
+    constructProperty_minimum_angle(-SimTK::Infinity);
+    constructProperty_maximum_angle(SimTK::Infinity);
+}
+
+//=============================================================================
+//  MocoFrameOrientationConstraint
+//=============================================================================
+
+MocoFrameOrientationConstraint::MocoFrameOrientationConstraint() 
+{
+    constructProperties();
+}
+void MocoFrameOrientationConstraint::initializeOnModelImpl(const Model& model, const MocoProblemInfo&) const
+{
+
+    int nFramePairs = getProperty_frame_pairs().size();
+    MocoConstraintInfo info;
+    std::vector<MocoBounds> bounds;
+
+    for (int i = 0; i < nFramePairs; ++i) {
+        const auto frame1_path = get_frame_pairs(i).get_frame1_path();
+        OPENSIM_THROW_IF(!model.hasComponent<Frame>(frame1_path), Exception,
+            "Could not find frame '{}'.", frame1_path);
+        auto& frame1 = model.getComponent<Frame>(frame1_path);
+        const auto frame2_path = get_frame_pairs(i).get_frame2_path();
+        OPENSIM_THROW_IF(!model.hasComponent<Frame>(frame2_path), Exception,
+            "Could not find frame '{}'.", frame2_path);
+        auto& frame2 = model.getComponent<Frame>(frame2_path);
+        m_frame_pairs.emplace_back(&frame1, &frame2);
+
+        const double& minimum = get_frame_pairs(i).get_minimum_angle();
+        const double& maximum = get_frame_pairs(i).get_maximum_angle();
+        
+        OPENSIM_THROW_IF(minimum > maximum, Exception,
+            "Expected the minimum angle for this frame pair "
+            "to be less than or equal to the maximum angle, "
+            "but they are {} and {}, respectively.",
+            minimum, maximum);
+        bounds.emplace_back(minimum, maximum);
+    }
+  
+
+    setNumEquations(nFramePairs);
+    info.setBounds(bounds);
+    const_cast<MocoFrameOrientationConstraint*>(this)->setConstraintInfo(info);
+
+}
+
+void MocoFrameOrientationConstraint::calcPathConstraintErrorsImpl(const SimTK::State& state, SimTK::Vector& errors) const
+{
+    int iconstr = 0;
+
+    getModel().realizePosition(state);
+
+    for (const auto& frame_pair : m_frame_pairs) {
+        const auto& frame1_rotation = frame_pair.first->getRotationInGround(state);
+        const auto& frame2_rotation = frame_pair.second->getRotationInGround(state);
+        auto coordaxis = SimTK::CoordinateAxis(2);											//set the number as 2 to represent ZAxis

```

Comment:
We don't want to hard-code a specific coordinate axis for the constraint, it would be better to provide users an option for which coordinate axis they want via a property. Alternatively, you could constrain all axes of the two frame with one scalar constraint by converting the orientation error to an angle-axis representation ([like we do in MocoOrientationTrackingGoal](https://github.com/opensim-org/opensim-core/blob/48ba9a0792c30b66bb2b97081f2546d1f627c2bc/OpenSim/Moco/MocoGoal/MocoOrientationTrackingGoal.cpp#L201)). 

Best would be provide an option to do either: if they provide a coordinate axis value, then you only minimize the error for that axis, if not then you use the angle-axis approach. Let me know if this makes sense to you.

Reply:
Hi @nickbianco ,

> Best would be provide an option to do either: if they provide a coordinate axis value, then you only minimize the error for that axis, if not then you use the angle-axis approach. Let me know if this makes sense to you.

This has been done, and the coordinate axis has been set as an optional property.